aya/src/programs/extension.rs

```diff
@@ -0,0 +1,178 @@
+use std::{
+    convert::TryFrom,
+    os::unix::prelude::{AsRawFd, RawFd},
+};
+
+use object::Endianness;
+
+use crate::{
+    generated::{bpf_attach_type::BPF_CGROUP_INET_INGRESS, bpf_prog_type::BPF_PROG_TYPE_EXT},
+    obj::btf::BtfKind,
+    programs::{load_program, LinkRef, ProgramData, ProgramError},
+    sys::{self, bpf_link_create},
+    Btf,
+};
+
+use super::{FdLink, Program};
+
+/// A program used to extend existing BPF programs
+///
+/// [`Extension`] programs can be loaded to replace a global
+/// function in a program that has already been loaded.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 5.9
+///
+/// # Examples
+///
+/// ```no_run
+/// use aya::{BpfLoader, programs::{Extension, LinkRef}};
+/// use std::convert::TryInto;
+///
+/// let mut bpf = BpfLoader::new().load_file("extension.o")?;
+/// let ext: &mut Extension = bpf.program_mut("extension").unwrap().try_into()?;
+///
+/// let root_prog = LinkRef::from_pinned_path("/sys/fs/bpf/root_prog")?;
+/// let link_fd = root_prog.fd().unwrap();
+/// ext.load(link_fd, "function_to_replace".to_string())?;
+/// ext.attach()?;
+/// Ok::<(), aya::BpfError>(())
+/// ```
+#[derive(Debug)]
+#[doc(alias = "BPF_PROG_TYPE_EXT")]
+pub struct Extension {
+    pub(crate) data: ProgramData,
+}
+
+impl Extension {
+    /// Loads the program inside the kernel.
+    ///
+    /// See also [`Program::load`](crate::programs::Program::load).
+    pub fn load<T: AsRawFd>(&mut self, program: T, func_name: String) -> Result<(), ProgramError> {
+        let target_prog_fd = program.as_raw_fd();
+
+        let info = sys::bpf_obj_get_info_by_fd(target_prog_fd).map_err(|io_error| {
+            ProgramError::SyscallError {
+                call: "bpf_obj_get_info_by_fd".to_owned(),
+                io_error,
+            }
+        })?;
+        if info.btf_id == 0 {
```

Comment:
if this would indicate a kernel bug, make it an assert!()

Reply:
i've added a typed error to catch this. it's not a kernel bug as it's possible for a user to try and replace f unction in a loaded program which hasn't loaded BTF into the kernel.